source/CIS/cis-3.rst

```diff
@@ -0,0 +1,144 @@
+execute a specific action on its behalf.
+
+The interface provides an entrypoint for querying the smart contract about which
+entrypoints can be invoked with a sponsored transaction.
+The smart contract responds with either: not supported, or supported.
+
+Specification
+=============
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED",  "MAY", and "OPTIONAL" in this document are to be interpreted as described in :rfc:`2119`.
+
+Logged events
+-------------
+
+This standard does not introduce any additional events. Entrypoints invoked via the `permit` function using the sponsored
+transaction mechanism MUST emit the same events as if they were invoked
+by the sponsoree directly without the sponsored transaction mechanism.
+
+Contract function
+-----------------
+
+A smart contract implementing CIS-3 MUST export the functions: :ref:`CIS-3-functions-permit` and :ref:`CIS-3-functions-supportsPermit`.
+
+.. _CIS-3-functions-permit:
+
+``permit``
+^^^^^^^^^^
+
+Verifies an ed25519 signature from a sponsoree and authorizes the sponsor to execute the logic of
+specific entrypoints on behalf of the sponsoree. The sponsored transaction mechanism replaces the
+authorization checks conducted on the `sender/invoker` variable with signature verification.
+
+Parameter
+~~~~~~~~~
+
+The parameter (``PermitParam``) contains a two-level signature map and a signer account that created the signature
+together with the message that was signed.
+
+.. note::
+
+    The CIS3 standard supports multi-sig accounts which is the purpose of the two-level signature map. A basic account (no multi-sig account) SHOULD have its signature at the key 0 in both maps.
+
+The message (``PermitMessage``) contains a contract_address (``ContractAddress``), entry_point (``OwnedEntrypointName``), nonce (``u64``), timestamp (``Timestamp``), and the payload (``PermitPayload``).
+This message structure enables the sponsoree the authorize the sponsor to act on its behalf in the given scope.
+
+The payload (``PermitPayload``) is serialized as: The first byte to distinguish the entrypoint followed by the parameter intended for that entrypoint::
+
+  PermitPayload  ::= (0: Byte) (Entrypoint_1_Parameter) // First entrypoint supported
+                   | (1: Byte) (Entrypoint_2_Parameter) // Second entrypoint supported
+                   | (2: Byte) (Entrypoint_3_Parameter) // Third entrypoint supported
+                   ...
+
+  PermitMessage ::= (contract_address: ContractAddress) (entry_point: OwnedEntrypointName) (nonce: u64) (timestamp: Timestamp) (payload: PermitPayload)
+
+  PermitParam ::= (signature: BTreeMap<u8, BTreeMap<u8, SignatureEd25519>>) (signer: AccountAddress) (message: PermitMessage)
+
+.. note::
+
+    The PermitPayload is an enum and the CIS-3 standard specifies that the first byte is used to distinguish between the different entrypoints.
```

Comment:
Since different entrypoints take different payloads the only real alternative is that the PermitPayload would start with entrypoint name. And that can then be used to distinguish how to parse the rest.

Reply:
Right, so we need to support tagged enum with a tag more than one byte in the schema.

Reply:
That's not entirely good either.

The enum has specific serialization which will not match that of an entrypoint name.

Reply:
We would have to support String tags I think. That would be the most natural in this case.

Reply:
I am not sure this tag is really a problem. The permit entrypoiint is meant to be called by an account, and has special parameters anyhow. The schema records the tags.

If we wanted to put the tags into the contract we could revise the `supportsPermit` entrypoint to return some metadata for the entrypoint it supports. That metadata is the tag that must then be used when calling `permit` with that entrypoint.

Reply:
In practice, when someone interacts with the smart contract, just the `payload (inputParameter to the function called via the sponsored transaction mechanism)` and the `schema` needs to be provided and the enum tag is sorted out for them by our tooling/helperFunctions. In that sense, I feel we can keep it as it is since no raw/manual enum tag derivation or interactions are needed.

https://github.com/Concordium/concordium-dapp-examples/blob/e6f0be6cb406b585602786f1cd19c39385d8e14b/sponsoredTransactions/front-end/src/SponsoredTransactions.tsx#L220

Reply:
Yes, right now people would probably rely on the schema, but if we want to add CIS-3 helper functions to something like the Rust SDK these would have to adjust the serialization based on each contract and would only be able to if that contract provided the missing information in the schema.

Reply:
To avoid having `random` tag assignment in the `PermitPayload` enum, we could standardize it so that the enum should be ordered based on the `entrypoint names`. 

In practice, a developer should execute `cargo concordium build -e` and use the order that the function names are printed to define the `PermitPayload` enum (The function entry point names are in a BTreeMap and this map prints out the functions in an orderly manner, where it is assumed that this ordering will never ever change in the future). 

e.g. `Transfer` has to be tag 0 and `UpdateOperator` has to have tag 1 and not the other way around.
```
enum PermitPayload {
    Transfer(TransferParameter),
    UpdateOperator(UpdateOperatorParams),
}
```
Even with this rule, any helper function (in e.g. the Rust SDK) would likely need to have access to the schema to serialize the `TransferParameter`/`UpdateOperatorParams` to provide some meaningful functionality, in my opinion.

Reply:
Another option, which should be fine in practice since the transaction is not really shown to the user, is to do
```

/// Part of the parameter type for the contract function `permit`.
/// Specifies the message that is signed.
#[derive(SchemaType, Serialize)]
struct PermitMessage {
    /// The contract_address that the signature is intended for.
    contract_address: ContractAddress,
    /// The entry_point that the signature is intended for.
    entry_point:      OwnedEntrypointName,
    /// A nonce to prevent replay attacks.
    nonce:            u64,
    /// A timestamp to make signatures expire.
    timestamp:        Timestamp,
    /// The serialized payload that should be forwarded to either the `transfer` or the
    /// `updateOperator` function.
    #[size_length = 2]
    payload:           Vec<u8>
}
```

That would avoid the tags at the cost of 2 extra bytes.

Reply:
Another solution that trades readability of the inner transaction via the schema is to use

```rust
/// Part of the parameter type for the contract function `permit`.
/// Specifies the message that is signed.
#[derive(SchemaType, Serialize)]
struct PermitMessage {
    /// The contract_address that the signature is intended for.
    contract_address: ContractAddress,
    /// The entry_point that the signature is intended for.
    entry_point:      OwnedEntrypointName,
    /// A nonce to prevent replay attacks.
    nonce:            u64,
    /// A timestamp to make signatures expire.
    timestamp:        Timestamp,
    /// The serialized payload that should be forwarded to either the `transfer` or the
    /// `updateOperator` function.
    #[size_length = 2]
    payload:           Vec<u8>
}
```

But if we have access to the schema of the entrypoint we can fully decode this parameter as well.

This looks like a good tradeoff.